### PR TITLE
feat(tech-insights): creates new endpoint for fact recalculation

### PR DESCRIPTION
**Pull request by sonikro**
## Hey, I just made a Pull Request!

Closes https://github.com/backstage/backstage/issues/18286

This PR adds a new endpoint in the tech-insights router to allow users to trigger fact recalculations for a single entity, instead of being forced to wait for the scheduled job to run.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by sonikro** [comment]
@Xantier , before I continue with more development and tests, I'd like to get your input about the direction of this implementation.

This is currently working, and if I call the new endpoint, it will trigger all FactRetrievers with an entityFilter for that single entity

image
image

Any concerns with this implementation?

**Comment by Xantier** [comment]
> This is currently working, and if I call the new endpoint, it will trigger all FactRetrievers with an entityFilter for that single entity

It might be beneficial to add a possibility to (optionallly) reference individual fact retrievers here. Some of them might be very high effort and wouldn't need to be necessarily run with a btn click/API call

**Comment by sonikro** [comment]
Can this PR be re-opened ? I'm finding some time to wrap it up. Also made some new commits to add the filter by FactRetrieverID. Now working on unit tests

**Comment by awanlin** [comment]
Hi @sonikro, I've re-opened it 👍

**Comment by sonikro** [comment]
> Hi @sonikro, I've re-opened it 👍

Thank you. I'm working on some tests right now to make sure everything works. I'll making the new method receive a list of factRetrieverIds, so people can either recalculate all facts of all factretrievers (if they dont specify any ID), or just recalculate facts for some specific fact retrievers

**Comment by sonikro** [comment]
@Xantier , can you take a look again? I added the ability to filter which factRetrievers you want to recalculate, and also added tests for the router and the new method in the DefaultEngine

**Comment by sonikro** [comment]
I'm a little bit concerned with the change in the **RouterOptions**. Technically the **factRetrieverEngine** is already being passed to the createRouter in all the examples, because of the spread object in here:

https://github.com/backstage/backstage/blob/aed0534408b8d543f90afb01e749c696c64a2efb/packages/backend/src/plugins/techInsights.ts#L102

So, just adding the engine to the interface just works.

However, because people that are consuming the plugin can decide how they want to call the **createRouter**, they might not be doing the spread and might not be passing the factRetrieverEngine. So technically this _could_ be a breaking change....

Thoughts?

File: plugins/tech-insights-backend/api-report.md [hunk 1]
```diff
@@ -100,6 +111,7 @@ export interface RouterOptions<
 > {
   config: Config;
   factChecker?: FactChecker<CheckType, CheckResultType>;
+  factRetrieverEngine: FactRetrieverEngine;
```
Comment (vinzscam) [inline plugins/tech-insights-backend/api-report.md:114]: yeah as you mentioned in your comment this is a breaking change.
Perhaps you can make it optional and throw some warnings if people try to use the new endpoint and `factRetrieverEngine` hasn't been passed?

File: plugins/tech-insights-backend/src/service/router.ts [hunk 2]
```diff
@@ -191,6 +198,18 @@ export async function createRouter<
     );
   });
 
+  router.post('/facts/:namespace/:kind/:name', async (req, res) => {
+    const { namespace, kind, name } = req.params;
+    const { factRetrieverIds }: { factRetrieverIds: string[] } = req.body;
```
Comment (vinzscam) [inline plugins/tech-insights-backend/src/service/router.ts:203]: let's add some validation to verify that `factRetrieverIds` is actually a `string[]`

File: plugins/tech-insights-backend/src/service/router.ts [hunk 3]
```diff
@@ -191,6 +198,18 @@ export async function createRouter<
...
+    await factRetrieverEngine.recalculateFactsForComponent?.({
+      kind,
+      name,
+      namespace,
+      factRetrieverIds,
+    });
+    res.status(201).send();
```
Comment (vinzscam) [inline plugins/tech-insights-backend/src/service/router.ts:210]: I don't know the logic deeply so I might be wrong but according to the logic this looks like a `200` 😄

**Review by vinzscam — CHANGES_REQUESTED** [review]

File: plugins/tech-insights-backend/src/service/router.ts (hunk 3 above)
Comment (Xantier) [reply plugins/tech-insights-backend/src/service/router.ts:210]: It trigger recalculation which creates new facts on the `/facts` endpoint, but does not return a response. I believe 201 is correct here

**Pull request closed without merging**